mm2src/common/file_lock.rs

```diff
@@ -0,0 +1,98 @@
+use crate::now_ms;
+use std::path::Path;
+
+pub struct FileLock<T: AsRef<Path>> {
+    /// Filesystem path of the lock file.
+    lock_path: T,
+    /// The time in seconds after which an outdated lock file can be removed.
+    ttl_sec: u64,
+}
+
+/// Records timestamp to a file contents.
+fn touch(path: &dyn AsRef<Path>, timestamp: u64) -> Result<(), String> {
+    std::fs::write(path.as_ref(), timestamp.to_string()).map_err(|e| ERRL!("{:?}", e))
+}
+
+/// Attempts to read timestamp recorded to a file
+fn read_timestamp(path: &dyn AsRef<Path>) -> Result<u64, String> {
+    match std::fs::read_to_string(path) {
+        Ok(content) => content.parse().map_err(|e| ERRL!("{:?}", e)),
+        Err(e) => ERR!("{:?}", e)
+    }
+}
+
+impl<T: AsRef<Path>> FileLock<T> {
+    pub fn lock(lock_path: T, ttl_sec: u64) -> Result<Option<FileLock<T>>, String> {
+        match std::fs::OpenOptions::new().write(true).create_new(true).open(lock_path.as_ref()) {
+            Ok(_) => {
+                let file_lock = FileLock { lock_path, ttl_sec };
+                try_s!(file_lock.touch());
+                Ok(Some(file_lock))
+            },
+            Err(ref ie) if ie.kind() == std::io::ErrorKind::AlreadyExists => {
+                // See if the existing lock is old enough to be discarded.
+                let lm = match read_timestamp(&lock_path) {
+                    Ok(lm) => lm,
+                    Err(ie) => return ERR!("Error checking {:?}: {}", lock_path.as_ref(), ie)
```

Comment:
If the filesystem is full, or if the process is stopped before the file is written (or if the computer is restarted and only a part of operations is retained in filesystem journal) then it's possible that the file will be empty. The `lock` function should account for this, maybe falling back to a file's last-modified metadata. Otherwise it's possible for lock to be stuck.

Reply:
Thanks! You've reminded me that I had initial idea to treat empty (or broken) file as expired lock, the idea got lost somewhere during implementation 🙂 Will refactor.